doc/_build/html/_sources/settings.rst.txt
========
Settings
========

In *Jok3r*, **settings are fully and easily customizable**. This is one of
the goal of the tool: **Make it evolve quickly**. This page explains how settings
files stored in the ``settings/`` directory work and can be edited.

Toolbox Settings
================
Toolbox settings are stored in the file ``settings/toolbox.conf``. This is where all tools
used by *Jok3r* are referenced and configured.

As an example, here is a snippet from the file:

.. code-block:: text

    [...]

    [patator]
    name           = patator
    description    = Multi-purpose brute-forcer, with a modular design and a flexible usage
    target_service = multi
    install        = git clone https://github.com/lanjelot/patator.git .
    update         = git pull
    check_command  = python2.7 patator.py -h

    [clusterd]
    name           = clusterd
    description    = Application server attack toolkit (JBoss, ColdFusion, Weblogic, Tomcat, Railo, Axis2, Glassfish)
    target_service = http
    install        = git clone https://github.com/hatRiot/clusterd.git . && sudo pip2 install -r requirements.txt
    update         = git pull && sudo pip2 install -r requirements.txt
    check_command  = python2.7 ./clusterd.py -h

    [...]

Format of this configuration file is pretty straightforward. For each tool, 
a section is created using the syntax ``[toolname]`` and the following options
can/must be specified:

* **name (mandatory)**: The name of the tool as it will be displayed. Authorized charset 
  is ``[a-z0-9_-]``.
* **description (mandatory)**: A Short text describing the tool.
* **target_service (mandatory)**: Service that can be targeted using this tool 
  (e.g. http for *Nikto*). For services such as *Nmap*, *Metasploit* and so on, that can
  be used to target several kinds of service, use the special service name "multi".
* **install (optional)**: Command-line to use in order to install the tool. It supports 
  the use of some tags (See `Tags for Commands`_). It is considered
  as optional because *Jok3r* allows to insert in the toolbox some tools that are not
  directly handled by it; it is for example the case for *Nmap* and *Metasploit* by default. 
  But note that if you don't supply installation command, you will not be able to control
  the installation of the tool from *Jok3r* and it is thus not advised to do so.
* **update (optional)**: Command-line to use in order to update the tool. Basically,
  take the same consideration as with the *install* option.
* **check_command (optional)**: Command-line to use in order to check for a correct install.
  Usually, it just consists in running the tool without any argument or with the standard
  ``-h`` option to see if everything seems working well after a fresh install (no 
  dependencies errors or such). This option can be omitted.


Services Settings
=================
For each service, settings are stored in file ``settings/<service>.conf``.
There is **one configuration file per service** which contains global parameters and then
the settings related to all security checks (for this service).

Global settings for a Service
-----------------------------
As an example, here is the beginning of the configuration file for *HTTP*, i.e. 
``settings/http.conf``:

.. code-block:: text
    
    [config]
    default_port = 80
    protocol     = tcp
    categories   = recon, vulnscan, exploit, bruteforce
    auth_types   = glassfish, jboss, jenkins, lotusdomino, tomcat, weblogic, websphere, wordpress, joomla, drupal, opencart, magento

    [specific_options]
    https    = boolean
    webdav   = boolean
    language = list
    cms      = list
    server   = list 

    [supported_list_options]
    supported_language = java, php, asp, angularjs, coldfusion
    supported_cms      = wordpress, joomla, drupal, mambo, silverstripe, vbulletin, magento, prestashop, liferay, opencart, dotnetnuke, django-cms, concrete5, punbb, moodle, cms-made-simple
    supported_server   = iis, glassfish, jboss, jenkins, tomcat, weblogic, lotusdomino

Every ``<service>.conf`` file begins with special following sections:

* ``[config]``: The basic configuration about the service. It contains the following options:

    * **default_port (mandatory)**: The default port number for the service.
    * **protocol (mandatory)**: The protocol (tcp or udp) for the service.
    * **categories (mandatory)**: List of different categories of checks supported for the service. 
      Authorized charset is ``[a-z0-9_-]``.
    * **auth_types (optional)**: List of authentication types that are supported for the service.
      Actually, only relevant for *HTTP* (where there are different possible authentications: *Tomcat, 
      JBoss, Wordpress, Joomla...*). Authorized charset is ``[a-z0-9_-]``.

* ``[specific_options]``: Contains the list of available context-specific options for the service.

    * **For each option, the name** (authorized charset is ``[a-z0-9_-]``) **is used as key** and
      the type as value. Supported types are:

        * ``boolean`` for boolean options. It is also possible to add the default value: 
          ``boolean:True`` means a boolean option which value is *True* if the option is not set.
          Note that ``boolean:False`` is redundant with ``boolean`` because *False* is the default.
        * ``list`` for options taking their value into a defined list (in ``[supported_list_options]``
          section, see below).
        * ``var`` for option of type variable.

* ``[supported_list_options]``: This section is used only if there is at least one context-specific
  option with the type ``list``. 

        * For each option of type ``list``, a key named **supported_<optionname>** is created and it 
          takes as value the list of authorized/supported values for the option.

.. note::
    For overall consistency, take care to use standard category names, among:

    * recon
    * vulnscan
    * exploit
    * bruteforce
    * postexploit


Security Checks for a Service
-----------------------------
For example, here are the settings of two checks as defined inside ``settings/http.conf``:

.. code-block:: text

    [check_jboss-deploy-shell]
    name        = jboss-deploy-shell
    category    = exploit
    description = Try to deploy shell on JBoss server (jmx-console, web-console, admin-console, JMXInvokerServlet)
    tool        = jexboss
    command_1   = python2.7 jexboss.py --auto-exploit --jboss -u [URL] --cmd whoami
    context_1   = { 'server': 'jboss', 'auth_status': NO_AUTH, 'auth_type': 'jboss' }
    command_2   = python2.7 jexboss.py --auto-exploit --jboss -u [URL] --jboss-login '[USERNAME]:[PASSWORD]' --cmd whoami
    context_2   = { 'server': 'jboss', 'auth_status': POST_AUTH, 'auth_type': 'jboss' }

    [check_web-path-bruteforce-targeted]
    name        = web-path-bruteforce-targeted
    category    = bruteforce
    description = Bruteforce web paths when language is known (extensions adapted) (use raft wordlist)
    tool        = dirsearch
    command_1   = python3 dirsearch.py -u [URL] -e jsp,java,do,txt,html,log -w [WORDLISTSDIR]/services/http/discovery/raft-large-directories.txt -f --exclude-status=400,404,500,000
    context_1   = { 'language': 'java' }
    command_2   = python3 dirsearch.py -u [URL] -e php,txt,html,log -w [WORDLISTSDIR]/services/http/discovery/raft-large-directories.txt -f --exclude-status=400,404,500,000
    context_2   = { 'language': 'php' }
    command_3   = python3 dirsearch.py -u [URL] -e asp,aspx,txt,html,log -w [WORDLISTSDIR]/services/http/discovery/raft-large-directories.txt -f --exclude-status=400,404,500,000
    context_3   = { 'language': 'asp' }
    command_4   = python3 dirsearch.py -u [URL] -e js,txt,html,log -w [WORDLISTSDIR]/services/http/discovery/raft-large-directories.txt -f --exclude-status=400,404,500,000
    context_4   = { 'language': 'angularjs' }
    command_5   = python3 dirsearch.py -u [URL] -e cfm,txt,html,log -w [WORDLISTSDIR]/services/http/discovery/raft-large-directories.txt -f --exclude-status=400,404,500,000
    context_5   = { 'language': 'coldfusion' }

Actually, each security check is defined under a section named ``[check_<check-name>]``
(authorized charset is ``[a-z0-9_-]``) by using the following options:

* **name (mandatory)**: The name of the tool as it will be displayed. Authorized 
  charset is ``[a-z0-9_-]``.
* **category (mandatory)**: Category inside which this check is classified. The name
  of the category must be in the list given in the option **categories** under the 
  section ``[config]`` at the beginning of the configuration file.
* **description (mandatory)**: Short text describing the check.
* **tool (mandatory)**: Name of the tool to use in this check. It must correspond
  exactly to the name which is given in ``toolbox.conf``.
* Each check is defined by one or several commands to run. For each command, you 
  should consider:

    * **command_<number> (mandatory)**: Command-line to run. It supports the use of multiple 
      tags (See `Tags for Commands`_)   
    * **context_<number> (optional)**: Context that must be met to run the corresponding
      command (See `Context Syntax`_)

* **postrun (optional)**: Name of a method from :ref:`smartmodules` to run after each/the
  command.


Tags for Commands
=================
Commands in settings supports the use of several tags. At runtime, they are replaced
by the correct values.

**For the commands in settings "install" and "update" in** ``settings/toolbox.conf``:

* ``[TOOLBOXDIR]``: Absolute path of toolbox directory.

**For the commands in setting "command_<number>" of security checks in** ``settings/<service>.conf``:

* ``[IP]``: The target IP address.
* ``[URL]``: The target URL (when target service is *HTTP*).
* ``[HOST]``: The target host (if not provided or reverse DNS lookup does not returns anything,
  IP address is used instead).
* ``[PORT]``: Target port number.
* ``[PROTOCOL]``: Protocol to use, either TCP or UDP.
* ``[SERVICE]``: Target service name.
* ``[WEBSHELLSDIR]``: Absolute path of directory storing webshells (useful for some exploits
  against *HTTP* services).
* ``[WORDLISTSDIR]``: Absolute path of directory storing wordlists.
* ``[USERNAME]``: Username for target from credentials store. This tag is supported for *Context* 
  (See `Context Syntax`_) with ``auth_status=USER_ONLY`` (only valid username is known), 
  or ``auth_status=POST_AUTH`` (valid username+password are known). If there are several
  usernames for the target in credentials store, the command is run for every username.
* ``[PASSWORD]``: Password for target from credentials store. This tag is supported for 
  *Context* with ``auth_status=POST_AUTH`` (valid username+password are known).
* ``[LOCALIP]``: Local IP address (might be useful for some exploits that perform reverse
  connection).


Specific tags depending on context-specific options are also supported by the service.
    
* **For context-specific option of type boolean**:
  ``[OPTION_NAME true="text to use if option is True"]``

* **For context-specific option of type list**:
  ``[OPTION_NAME element1="val1" element2="val2" ... default="default val" ]`` 
  If the option has the value "element1",
  - respectively "element2" - the tag will be replaced by "val1" - respectively "val2".
  If the value of the option does not match anything specified in the tag, the tag
  will be replaced by "default val" ("default" parameter optional).

* **For context-specific option of type var**:
  ``[OPTION_NAME set="text _VAR_ text" default="default text"]``

    * If variable is set, it is replaced by the text into "set" parameter 
      with "_VAR_" replaced by variable's value.
    * Otherwise, it is replaced by the text into "default" parameter if existing
      (optional parameter).


Context Syntax
==============
For each command in setting "command_<number>" of security checks in ``settings/<service>.conf``,
it is possible to specify a context. **A context defines the required conditions to run 
the command**.

For the setting "command_<number>", a context can be defined in setting "context_<number>"
(<number> must match). The context is defined using a Python dictionary.

Here is an example for a command in a security check against *HTTP*: 

.. code-block:: text

    command_2   = python2.7 jexboss.py --auto-exploit --jboss -u [URL] --jboss-login '[USERNAME]:[PASSWORD]' --cmd whoami
    context_2   = { 'server': 'jboss', 'auth_status': POST_AUTH, 'auth_type': 'jboss' }

The value of "context_2" means that "command_2" must be run **if and only if the following 
conditions are met**:

* The context-specific option ``server == 'jboss'``, i.e. the target *HTTP* service is using
  *JBoss* server.
* Valid credentials for *JBoss* on the targeted service are present in the credentials store in
  the database.

More generally, **conditions that can be defined in context are**:

* **Values of context-specific options**,
* **Authentication status on the target** via the key ``auth_status`` that can take either 
  of the following values:
  
    * ``NO_AUTH``: No credentials are known,
    * ``USER_ONLY``: At least one username is known,
    * ``POST_AUTH``: Valid credentials (username+password) are known,
    * ``None``: Any status.

.. warning::
    For *HTTP* only, if ``auth_status`` is used to define a context, it is mandatory to 
    specify for which kind of authentication does that must apply via the key ``auth_type``
    (in the previous example, it was for "jboss").

CHANGELOG.rst
=========
Changelog
=========
**v3.0 BETA 3 Arkham Subversion** 2020-07-21
	*Add RPC service and tool (disabled by default)
	*Add initial scheme for F5 BIG IP vulnerability scanner and exploiter
	*Add/fix/customize some bits here and there
	*First pre-release v0.1

**v3.0 BETA 3** 2020-07-18
	* Add waf-checker and Pentest1 to attacks profiles
	* Fix several minor bugs and customizations
	* Kali GPG key updated

**v3.0 BETA 2** 2019-07-09
	* Fix several bugs after testings
	* Add/fix/improve matchstrings
	* Add/fix/improve checks
	* Changes in database structure to store more info related to hosts (OS, device type, vendor)
	* Update on HTML reporting templates (more user-friendly)

**v3.0 BETA** 2019-04-23
	* Support for products detection based on checks outputs (e.g web-server, web-appserver, ftp-server...)
	* Support for vulnerabilities detection based on checks outputs (still experimental !)
	* Database structure updated to be able to work with products and vulnerabilities
	* Add search feature in local database
	* Full redesign of smartmodules aimed at updating the context (based on regex for creds, options, products & vulns)
	* Improve web technologies detection at beginning of attacks against HTTP services
	* Add many checks/tools
	* Add CVE lookup when product name & version have been detected
	* Add HTML reporting (with screenshot support for HTTP services)
	* Improve web content discovery (smart choice of wordlists based on the context)
	* Add support for custom attack profiles (pre-selection of checks)
	* Targets initialization optimization
	* Add support for --userlist and --passlist to use custom wordlists for bruteforce checks
	* Improve context requirement feature
	* Code refactoring and code docstrings
	* Fix many bugs

**v2.0** 2018-10-19
	* Everything rebuilt from scratch
	* Python3 compatibility
	* Local database to store targets and results from checks
	* Context awareness to run only relevant checks against targets
	* Post processing of checks outputs possible, in order to update context
	* Dockerization
	* Add lots of checks based on various tools
	* Add supported services
	* ...and many more

**v1.1**
    * Many bug fixes
    * Add lots of services/tools
    * Improve wordlists
    * Support for "multi-services" tools (tools that can be used differently for targeting different services)
    * Change a bit of settings logic: Now, install status and last update time are stored into an external file called "_install_status.conf"

**v1.0** 2017-03-10 
	* First release

lib/core/Config.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
###
### Core > Config
###
import colored
import os

from lib.core.Constants import *
from lib._version import __version__


#----------------------------------------------------------------------------------------
# Banner/Help

BANNER = colored.stylize("""

                          `-:/++++/:-.    .-:/++++/:-`                                    
                        .:ohdddmmmmdd.\  /.dddmmmmdddho:.                                
                      `:ydmmmmmmmmmmmmm\/mmmmmmmmmmmmmmdy:`                         
                     `+dmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmd+`                     
                    +dyo+++oshmmmmmmmmmmmmmmmmmmmmhso+++oyd+                    
                  -+-         .dmmmmmmmmmmmmmmmmd.         -+-                  
                 ``           `dmmmmmmmmmmmmmmmmd`           ``                 
                              `dmmmmmmmmmmmmmmmmd`                              
                              `ymmmmmmmmmmmmmmmmy`                              
                                .+dmmmmmmmmmmd+.                                
                                   /dmmmmmmd/                                   
                                    `odmmdo`                                    
                                      .hh.                                      
                                                        
                                                                   
                       ██╗ ██████╗ ██╗  ██╗██████╗ ██████╗ 
                       ██║██╔═══██╗██║ ██╔╝╚════██╗██╔══██╗
                       ██║██║   ██║█████╔╝  █████╔╝██████╔╝
                  ██   ██║██║   ██║██╔═██╗  ╚═══██╗██╔══██╗
                  ╚█████╔╝╚██████╔╝██║  ██╗██████╔╝██║  ██║  v{version}
                   ╚════╝  ╚═════╝ ╚═╝  ╚═╝╚═════╝ ╚═╝  ╚═╝ 
                  
                  [ Network & Web Pentest Automation Framework ]

""".format(version=__version__), colored.fg('light_green') + colored.attr('bold'))

USAGE = """
python3 jok3r.py <command> [<args>]

Supported commands:
   toolbox    Manage the toolbox
   info       View supported services/options/checks
   db         Define missions scopes, keep tracks of targets & view attacks results
   attack     Run security checks against targets
   
"""

ATTACK_EXAMPLES = colored.stylize('Examples:', colored.attr('bold')) + """
  - Run all security checks against an URL in interactive mode (break before each check):
  python3 jok3r.py attack -t http://www.example.com/ 

  - Run all security checks against a MS-SQL service (without user interaction) and add results to the mission "mayhem" in db:
  python3 jok3r.py attack -t 192.168.1.42:1433 -s mssql --add2db mayhem --fast

  - Run only "recon" and "vulnscan" security checks against an FTP service and add results to the mission "mayhem" in db:
  python3 jok3r.py attack -t 192.168.1.142:21 -s ftp --cat-only recon,vulnscan --add2db mayhem

  - Run security checks against all FTP services running on 2121/tcp and all HTTP services from the mission "mayhem" in db:
  python3 jok3r.py attack -m mayhem -f "port=2121;service=ftp" -f "service=http"

  - Search for "easy wins" (critical vulns / easy to exploit) on all services registered in mission "mayhem" in db:
  python3 jok3r.py attack -m mayhem --profile red-team --fast
"""

DB_INTRO = """
The local database stores the missions, targets info & attacks results.
This shell allows for easy access to this database. New missions can be added and
scopes can be defined by importing new targets.
"""


#----------------------------------------------------------------------------------------
# Arguments Parsing Settings

ARGPARSE_MAX_HELP_POS    = 45
TARGET_FILTERS           = {
    'ip'       : FilterData.IP, 
    'host'     : FilterData.HOST,
    'port'     : FilterData.PORT, 
    'service'  : FilterData.SERVICE, 
    'url'      : FilterData.URL,
    'osfamily' : FilterData.OS_FAMILY,
    'banner'   : FilterData.BANNER,
}


#----------------------------------------------------------------------------------------
# Basic Settings

TOOL_BASEPATH      = os.path.dirname(os.path.realpath(__file__+'/../..'))
TOOLBOX_DIR        = TOOL_BASEPATH + '/toolbox'
DEFAULT_OUTPUT_DIR = 'output'
WEBSHELLS_DIR      = TOOL_BASEPATH + '/webshells'
WORDLISTS_DIR      = TOOL_BASEPATH + '/wordlists'
DB_FILE            = TOOL_BASEPATH + '/local.db'
DB_HIST_FILE       = TOOL_BASEPATH + '/.dbhistory'
REPORT_TPL_DIR     = TOOL_BASEPATH + '/lib/reporter/templates'
REPORT_PATH        = TOOL_BASEPATH + '/reports'
VIRTUALENVS_DIR    = TOOL_BASEPATH + '/toolbox/virtualenvs'

#----------------------------------------------------------------------------------------
# Display Settings

ATTACK_SUMMARY_TABLE_MAX_SIZE = 18

#----------------------------------------------------------------------------------------
# Settings Files

SETTINGS_DIR              = TOOL_BASEPATH + '/settings'
CONF_EXT                  = '.conf'
TOOLBOX_CONF_FILE         = 'toolbox'
INSTALL_STATUS_CONF_FILE  = '_install_status'
ATTACK_PROFILES_CONF_FILE = 'attack_profiles'
PREFIX_SECTION_CHECK      = 'check_'
MULTI_CONF                = 'multi'
MULTI_TOOLBOX_SUBDIR      = 'multi'

TOOL_OPTIONS = {
    MANDATORY: [
        'name',
        'description',
        'target_service',
    ],
    OPTIONAL: [
        'virtualenv',
        'install',
        'update',
        'check_command',
    ]
}

SERVICE_CHECKS_CONFIG_OPTIONS = {
    MANDATORY: [
        'default_port',
        'protocol',
        'categories',
    ],
    OPTIONAL: [
        'auth_types'
    ]
}

CHECK_OPTIONS = {
    MANDATORY: [
        'name',
        'category',
        'description',
        'tool',
        # command
    ],
    OPTIONAL: [
        'apikey',
    ]
}

OPTIONS_ENCRYTPED_PROTO = (
    'ftps',
    'https',
    'rmissl',
    'smtps',
    'telnets',
)

